File: backend/src/main/java/com/team01/backend/global/security/CustomAccessDeniedHandler.java
package com.team01.backend.global.security;

import com.team01.backend.global.response.ApiResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;
import tools.jackson.databind.ObjectMapper;

import java.io.IOException;

@Component
@RequiredArgsConstructor
public class CustomAccessDeniedHandler implements AccessDeniedHandler { // 인가 실패 관리

    // 필드 주입 -> 생성자 주입으로 수정
    private final ObjectMapper objectMapper;

    @Override
    public void handle(HttpServletRequest request,
                       HttpServletResponse response,
                       AccessDeniedException ex) throws IOException {

        response.setStatus(HttpServletResponse.SC_FORBIDDEN);
        response.setContentType("application/json;charset=UTF-8");

        ApiResponse<Void>apiResponse = ApiResponse.ofFailure("FORBIDDEN", "권한이 없습니다.");

        response.getWriter().write(
                objectMapper.writeValueAsString(apiResponse)
        );
    }
}

File: backend/src/main/java/com/team01/backend/global/security/CustomAccessDeniedHandler.kt
package com.team01.backend.global.security

import com.fasterxml.jackson.databind.ObjectMapper
import com.team01.backend.global.response.ApiResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component

@Component
class CustomAccessDeniedHandler(
    private val objectMapper: ObjectMapper
) : AccessDeniedHandler {  // 인가 실패 관리

    override fun handle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        ex: AccessDeniedException
    ) {
        response.status = HttpServletResponse.SC_FORBIDDEN
        response.contentType = "application/json;charset=UTF-8"

        val apiResponse = ApiResponse.ofFailure<Void>("FORBIDDEN", "권한이 없습니다.")

        response.writer.write(
            objectMapper.writeValueAsString(apiResponse)
        )
    }
}